integration-tests/src/scenario_utils.rs

```diff
@@ -0,0 +1,236 @@
+        let compiled_contract = CompiledContract {
+            path: path_sol.to_str().expect("path is not str").to_string(),
+            name: name.to_string(),
+            abi,
+            bin: bin.into_bytes().expect("bin"),
+            bin_runtime: bin_runtime.into_bytes().expect("bin_runtime"),
+        };
+
+        let mut path_json = path_sol.clone();
+        path_json.set_extension("json");
+        serde_json::to_writer(
+            &File::create(&path_json).expect("cannot create file"),
+            &compiled_contract,
+        )
+        .expect("cannot serialize json into file");
+
+        contracts.insert(name.to_string(), compiled_contract);
+    }
+
+    contracts
+}
+
+/// Depoloy contract
+pub async fn deploy_contract<M, T>(
+    wallet: &Arc<M>,
+    compiled: &CompiledContract,
+    args: T,
+) -> (u64, Contract<M>)
+where
+    M: Middleware,
+    T: Tokenize,
+{
+    let factory = ContractFactory::new(compiled.abi.clone(), compiled.bin.clone(), wallet.clone());
+    let (contract, receipt) = factory
+        .deploy(args)
+        .unwrap_or_else(|_| panic!("cannot deploy: {:?}", compiled.name))
+        .confirmations(0usize)
+        .send_with_receipt()
+        .await
+        .unwrap_or_else(|_| panic!("cannot deploy: {:?}", compiled.name));
+
+    (receipt.block_number.unwrap().as_u64(), contract)
+}
+
+/// An `sender` account sends ETH to multiple accounts
+pub async fn distribute_eth(
+    prov: &Provider<Http>,
+    sender: &WalletType,
+    wallets: &[WalletType],
+) -> u64 {
+    // Fund NUM_TXS wallets from wallet0
+    let mut block_num: u64;
+    let mut pending_txs = Vec::new();
+    let mut nonce = prov
+        .get_transaction_count(sender.address(), None)
+        .await
+        .expect("cannot get transaction_count");
+
+    for wallet in wallets[0..NUM_TXS].iter() {
+        let tx = TransactionRequest::new()
+            .to(wallet.address())
+            .nonce(nonce)
+            .value(WEI_IN_ETHER * 2u8) // send 2 ETH
+            .from(sender.address());
+        pending_txs.push(
+            wallets[0]
+                .send_transaction(tx, None)
+                .await
+                .expect("cannot send tx"),
+        );
+        nonce = nonce.checked_add(U256::one()).unwrap();
+    }
+
+    sleep(Duration::from_secs(5));
+    block_num = 0;
+    for tx in pending_txs.into_iter() {
+        let receipt = tx.await.expect("cannot confirm tx").unwrap();
+        if block_num == 0 {
+            block_num = receipt.block_number.unwrap().as_u64();
+        } else if block_num != receipt.block_number.unwrap().as_u64() {
+            panic!("The txs are not in a block")
+        }
+    }
+    block_num
+}
+
+/// TransferContext
+pub struct TransferContext {
+    /// Index of `from` wallet
+    pub from_i: usize,
+    /// Index of `to` wallet
+    pub to_i: usize,
+    /// Amount to transfer
+    pub amount: U256,
+    /// Expected result
+    pub expected_result: bool,
```

Comment:
Let's simplify the names

```suggestion
    /// Index of `from` wallet
    pub from: usize,
    /// Index of `to` wallet
    pub to: usize,
    /// Amount to transfer
    pub amount: U256,
    /// Expected result
    pub expected: bool,
```